File: c++/leetcode/symmetric-tree/recursive.cpp
/**
 * https://leetcode.com/problems/symmetric-tree/
 *
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        if (!root) {
            return true;
        }
        if (root->left && root->right) {
            return compareNodes(root->left, root->right);
        } else if (!root->left && !root->right) {
            return true;
        }

        return false;
    }
    bool compareNodes(TreeNode* left, TreeNode* right) {
        if (!left && !right) {
            return true;
        } else if (left && !right || !left && right) {
            return false;
        }

        return left->val == right->val && compareNodes(left->left, right->right) && compareNodes(left->right, right->left);
    }
};

File: go/leetcode/traverse-a-tree/binary-tree-pinorder-traversal/recursive.go
package binary_tree_inorder_traversal

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func inorderTraversal(root *TreeNode) []int {
	result := make([]int, 0)

	if root == nil {
		return result
	}

	if root.Left != nil {
		for _, v := range inorderTraversal(root.Left) {
			result = append(result, v)
		}
	}
	result = append(result, root.Val)
	if root.Right != nil {
		for _, v := range inorderTraversal(root.Right) {
			result = append(result, v)
		}
	}

	return result
}
